rust/kernel/allocator.rs

```diff
@@ -2,12 +2,44 @@
 
 //! Allocator support.
 
+use core::alloc::AllocError;
 use core::alloc::{GlobalAlloc, Layout};
 use core::ptr;
+use core::ptr::NonNull;
 
 use crate::bindings;
 
-struct KernelAllocator;
+pub(crate) struct KernelAllocator;
+
+impl KernelAllocator {
+    #[cfg(not(test))]
+    #[cfg(not(testlib))]
+    pub(crate) fn allocate_with_flags(
+        &self,
+        layout: Layout,
+        flags: bindings::gfp_t,
+    ) -> Result<NonNull<[u8]>, AllocError> {
+        // `krealloc()` is used instead of `kmalloc()` because the latter is
+        // an inline function and cannot be bound to as a result.
+        let mem = unsafe { bindings::krealloc(ptr::null(), layout.size(), flags) as *mut u8 };
```

Comment:
This may not be related to your PR, but it's legal to have a layout whose `align` is greater than the `size`:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8440766fc0244b0b6b5327a969b176d7

So I think we need:

* either mark this function as `unsafe` and make sure "size>=align" is the safety requirement or,
* call a `Layout::pad_to_align` here to generate a "safe" `Layout`

Reply:
Doesn't `krealloc` guarantee an alignment of `ARCH_KMALLOC_MINALIGN`?

Reply:
> Doesn't `krealloc` guarantee an alignment of `ARCH_KMALLOC_MINALIGN`?

`ARCH_KMALLOC_MINALIGN` is a constant, for example on arch whose `ARCH_DMA_MINALIGN` <= 8, it's `sizeof (unsiged long long)`. What if a user passes a layout with size = 32 and alignment = 4096? Some virtualization may establish communication between guests and host via a shared page, but only the first few bytes are used.

Reply:
Well, in that case we are screwed. There are no allocation APIs with explicit alignment params in the kernel, so there are no ways that we can do allocation for a layout, say, with size = 192 and alignment = 64.

Reply:
> Well, in that case we are screwed. There are no allocation APIs with explicit alignment params in the kernel, so there are no ways that we can do allocation for a layout, say, with size = 192 and alignment = 64.

`kmalloc` and itself will "align to the size" when the size is power of two:

https://github.com/Rust-for-Linux/linux/blob/rust/include/linux/slab.h#L541 , so as long as we allocate with 256, it's fine.

Hmm... but I don't know what if the size = 260 and the alignment = 64. IOW, if we kmalloc(320), what the alignment we can get.

Reply:
> There are no allocation APIs with explicit alignment params in the kernel, so there are no ways that we can do allocation for a layout, say, with size = 192 and alignment = 64.

You can overallocate, align and write a header just before the returned pointer to allow getting the original allocated pointer to free. This is what libstd does on windows: https://github.com/rust-lang/rust/blob/efb171e2350de2bec6dd1f035b99bc00535c1c15/library/std/src/sys/windows/alloc.rs#L149

Reply:
> > There are no allocation APIs with explicit alignment params in the kernel, so there are no ways that we can do allocation for a layout, say, with size = 192 and alignment = 64.
> 
> You can overallocate, align and write a header just before the returned pointer to allow getting the original allocated pointer to free. This is what libstd does on windows: https://github.com/rust-lang/rust/blob/efb171e2350de2bec6dd1f035b99bc00535c1c15/library/std/src/sys/windows/alloc.rs#L149

Right, we can use that when size is greater than align and size is not a power of two (and align is greater than `ARCH_KMALLOC_MINALIGN`. If later on, we find that `kmalloc` already covers us on the alignment guarantee, it can be removed.

When size == align, we are covered by the current `kmalloc` guarantee.

And if size < align, we need to `pad_to_align`, then it's same as the "size == align" case.

Reply:
Updated PR to address comments. Based on https://github.com/rust-lang/rust/blob/efb171e2350de2bec6dd1f035b99bc00535c1c15/library/std/src/sys/windows/alloc.rs

Reply:
> OK, I will update the PR.
> 
> It seems like this consideration would also apply to `(KernelAllocator as GlobalAlloc)::alloc()`. Safety requirement does not mention this issue. Should we update the safety requirement or add the same kind of logic to the `alloc()` method?

I think so. But in the meanwhile let me drop an email to slab people and see whether we can get some extra alignment guarantee.

Reply:
> > OK, I will update the PR.
> > It seems like this consideration would also apply to `(KernelAllocator as GlobalAlloc)::alloc()`. Safety requirement does not mention this issue. Should we update the safety requirement or add the same kind of logic to the `alloc()` method?
> 
> I think so. But in the meanwhile let me drop an email to slab people and see whether we can get some extra alignment guarantee.

Hmm... read the slab code again. Looks to me that we don't have extra alignment guarantee when size is an odd multiple of power of two. So we indeed need to do the alignment on our own.

Reply:
FWIW, I have written a simple test for rust alloc to see if the current implementation provides the alignment requirement: https://github.com/Rust-for-Linux/linux/pull/817